app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,95 @@
+#
+# Handle signup or signin
+# from various oauth providers
+#
+# Heavily overlaps with Authentications controller
+#
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    create
+  end
+  def failure
+    flash[:error] = "Authentication failed."
+    redirect_to request.env['omniauth.origin'] || "/"
+  end
+
+  private
+
+  #
+  # Inspects the omniauth information
+  # and determines if we have an existing member 
+  # (to add authentication to)
+  # or if this is a new signup
+  #
+  # As a side affect, this method sets the @member_created
+  # variable
+  #
+  def member
+    @member_created = false
+
+    auth = request.env['omniauth.auth']
+
+    authentication = Authentication.where(provider: auth.provider, uid: auth.uid).first
+    if authentication && authentication.member.id.present?
+      member = authentication.member
+    end
+
+    member = Member.where(email: auth.info.email).first_or_create do |m|
+      m.email = auth.info.email
+      m.password = Devise.friendly_token[0,20]
+      m.tos_agreement = true
```

Comment:
We should split this out and have a workflow to ask for it specifically, where we can also get them to review and edit the login name to prevent collisions.